### PR TITLE
String and Array notation types for entity service populate Param

### PR DESCRIPTION
**Pull request by Bassel17**
### What does it do?

- Add string and array notation types 
- add populate as a possible param

### Why is it needed?

- To complete entity service typings

### CODE REVIEW

File: packages/core/strapi/lib/services/entity-service/types/params/populate.d.ts
Comment (innerdvations) [inline packages/core/strapi/lib/services/entity-service/types/params/populate.d.ts:37]: Are we adding advanced populate types in a later PR but before the target PR is merged? It feels like having something like `| Object` would work as an essentially untyped placeholder unti we get around to implementing it.

Comment (Bassel17) [reply packages/core/strapi/lib/services/entity-service/types/params/populate.d.ts:37]: yeah it is the next thing am working on, yeah I could add that

File: packages/core/strapi/lib/services/entity-service/types/params/populate.d.ts
Comment (Convly) [inline packages/core/strapi/lib/services/entity-service/types/params/populate.d.ts:22]: If you don't add a fallback to a regular string here, you won't be able to write

```ts
const populate: StringNotation<Common.UID.Schema> = 'foo';
```
 
 as foo is neither `'*'` nor `"${string},${string}"` or `never`

```suggestion
  | Utils.Guard.Never<Attribute.GetPopulatableKeys<TSchemaUID>, string>
```

File: packages/core/strapi/lib/services/entity-service/types/params/populate.d.ts
Comment (Convly) [inline packages/core/strapi/lib/services/entity-service/types/params/populate.d.ts:23]: We might want to add an `${string}.{string}` memeber too, as it's possible to do `populate='article.tags'`

File: packages/core/strapi/lib/services/entity-service/types/params/populate.d.ts
Comment (Convly) [reply packages/core/strapi/lib/services/entity-service/types/params/populate.d.ts:37]: Since nothing will be released before adding the object notation, I'm not sure it adds value. But no strong opinion either.

File: packages/core/strapi/lib/services/entity-service/types/params/populate.d.ts
Comment (Convly) [reply packages/core/strapi/lib/services/entity-service/types/params/populate.d.ts:22]: At the moment `Attribute.GetPopulatableKeys` returns a list of all relations, components, dynamic zones & media. However, do we consider morph relations (owner side) to be populated? (honestly not sure, might be better to test with the file.related attribute for instance)

Comment (Bassel17) [reply packages/core/strapi/lib/services/entity-service/types/params/populate.d.ts:22]: yeah but fallback incase of what ? you wouldn't want to populate something that doesn't exist

Comment (Convly) [reply packages/core/strapi/lib/services/entity-service/types/params/populate.d.ts:22]: Types not generated means any content type can have any field, so in this case, we should accept anything that can represent an attribute (=> a string)

File: packages/core/strapi/lib/services/entity-service/types/params/populate.d.ts
Comment (innerdvations) [inline packages/core/strapi/lib/services/entity-service/types/params/populate.d.ts:43]: This may be an issue with documentation rather than this implementation, but the documentation for both [rest API](https://docs.strapi.io/dev-docs/api/rest/populate-select#population) and [entity service](https://docs.strapi.io/dev-docs/api/entity-service/populate) only allow `WildCardNotation | ArrayNotation | ObjectNotation` and do not mention strings as an option anywhere.

File: packages/core/strapi/lib/services/entity-service/types/params/populate.d.ts
Comment (innerdvations) [inline packages/core/strapi/lib/services/entity-service/types/params/populate.d.ts:29]: This may be an issue with documentation rather than this implementation, but the documentation for both [rest API](https://docs.strapi.io/dev-docs/api/rest/populate-select#population) and [entity service](https://docs.strapi.io/dev-docs/api/entity-service/populate) do not mention either of these methods. They only say "an array of attribute names".

File: packages/core/strapi/lib/services/entity-service/types/params/populate.d.ts
Comment (innerdvations) [inline packages/core/strapi/lib/services/entity-service/types/params/populate.d.ts:40]: `['someField', '*']` will be treated as valid, which might be ok, but it's also not in the documentation. Should wildcard only be available when used by itself and not in an array?

File: packages/core/strapi/lib/services/entity-service/types/params/populate.d.ts
Comment (Convly) [reply packages/core/strapi/lib/services/entity-service/types/params/populate.d.ts:29]: Same answer as above :eyes:

**Comment by Convly** [comment]
> Coincidentally, this seems like it relates to the discussions in #16507
> 
> We might want to watch that and see what decisions are made about what values are valid

This shouldn't have an impact TBH, we're not changing which format is accepted, just how we sanitize them in some places (e.g. controllers or services)

File: packages/core/strapi/lib/services/entity-service/types/params/populate.d.ts
Comment (Convly) [reply packages/core/strapi/lib/services/entity-service/types/params/populate.d.ts:40]: Indeed, looking at the implementation, we're not really trying to couple both the wildcard & regular string. We could indeed exclude the wildcard from the array notation

Comment (Bassel17) [reply packages/core/strapi/lib/services/entity-service/types/params/populate.d.ts:40]: I mean it's possible and doesn't really throw an error

Comment (Bassel17) [reply packages/core/strapi/lib/services/entity-service/types/params/populate.d.ts:40]: we can't compeletely discard it from the array notation this ['*'] is still equivalent to this '*'

File: packages/core/strapi/lib/services/entity-service/types/params/populate.d.ts
Comment (Convly) [inline packages/core/strapi/lib/services/entity-service/types/params/populate.d.ts:29]:
```suggestion
  | `${PopulatableKeys<TSchemaUID>}.${string}`;
```

**Review by Convly — CHANGES_REQUESTED** [review]

**Review by Convly — COMMENTED** [review]
Still not 100% sure about what to do with the wildcard in the arrays situation but the rest looks nice. Waiting for a final decision on that & LGTM

File: packages/core/strapi/lib/services/entity-service/types/params/populate.d.ts
Comment (innerdvations) [reply packages/core/strapi/lib/services/entity-service/types/params/populate.d.ts:43]: Ok, so just a lack of documentation on those pages. I knew that the code supported it, it was a huge pain with the sanitization, I was curious if we could potentially drop any parts of it 😆

Pinging @pwizla in case we want to add simple attribute name strings and compound strings (like `"attribute.attribute"`) to the main documentation instead of only appearing suddenly in an advanced example.

File: packages/core/strapi/lib/services/entity-service/types/params/populate.d.ts
Comment (innerdvations) [reply packages/core/strapi/lib/services/entity-service/types/params/populate.d.ts:40]: I say go with however it's currently implemented. If `['someField', '*']` works to populate everything, this is fine.

**Review by innerdvations — APPROVED** [review]
LGTM!

**Review by christiancp100 — APPROVED** [review]
LGTM 🍝

**Review by Convly — APPROVED** [review]
I've QAed it & it works well. LGTM

**Pull request merged**
